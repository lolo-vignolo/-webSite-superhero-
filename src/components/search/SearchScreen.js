import { useMemo } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string'

import { getHeroByName } from '../../selectors/getHeroByName';

import { useForm } from '../../hooks/FromHook';
import { HeroCard } from '../heroes/HeroCard';


export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);
    
    const [ formValues, handleInputChange ] = useForm({
        searchText: q,
    });

    const { searchText } = formValues;

    const heroesFileted = useMemo( () => getHeroByName(q), [q] );


    const handleSearch = (e) => {
        e.preventDefault();
        navigate(`?q=${ searchText }`)
    }


    return (
        <>
            <h1>Búsquedas</h1>
            <hr />

            <div className="row">

                <div className="col-5">
                    <h4>Buscar</h4>
                    <hr />

                    <form onSubmit={ handleSearch }>
                        <input 
                            type="text"
                            placeholder="Buscar un héroe"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={ searchText }
                            onChange={ handleInputChange }
                        />


                        <button 
                            className="btn btn-outline-primary mt-1"
                            type="submit">
                            Buscar...
                        </button>

                    </form>


                </div>

                <div className="col-7">
                    <h4>Resultados</h4>
                    <hr />

                    {
                        (q === '')
                            ? <div className="alert alert-info"> Buscar un héroe </div>
                            : ( heroesFileted.length === 0 ) 
                                && <div className="alert alert-danger"> No hay resultados: { q } </div>
                    }


                    {
                        heroesFileted.map(hero => (
                            <HeroCard 
                                key={ hero.id }
                                { ...hero }
                            />
                        ))
                    }


                </div>

            </div>

        </>
    )
}


// import React, { useMemo } from 'react'
// import { useLocation } from 'react-router';
// import { useForm } from '../../hooks/FromHook';
// import { HeroCard } from '../heroes/HeroCard';
// import queryString from "query-string";
// import { getHeroByName } from '../../selectors/getHeroByName';

// export const SearchScreen = ({history}) => {
//     const location = useLocation();
//     const {q = ""} = queryString.parse(location.search);
   


//     const [values, handleImputChange,] = useForm({ 
//         superhero: q , 
//     });
    
//     const {superhero} = values
//     //superhero es el value del imput

//     const heroesFiltered = useMemo(() => getHeroByName (q), [q])
    

//     const handleSearch = (event)=> {
//         event.preventDefault();
//         history.push(`?q=${superhero}`);
//     }

//     return (
//         <div>
//             <h1>Search Screen </h1>
//             <hr />
//             <div className = " row">

//                 <div className = "col-5">
//                     <h4> Search form</h4>
//                     <hr />

//                     <form onSubmit = {handleSearch} >
//                        <input 
//                        type = "text"
//                        name = "superhero"
//                        value = {superhero}
//                        placeholder = "Find your hero"
//                        className = " form-control"
//                        onChange = {handleImputChange}
//                        />
//                         <button
//                         type = "submit"
//                         className = "btn m-1 btn-outline-primary btn-block"
//                         >
//                             Search...
//                         </button>

//                     </form>
//                 </div>

//                 <div className = "col-7">

//                 <h4> Results</h4>
//                 <hr />

//                 {
//                     (q === "") &&
//                     <div className = "alert alert-info"> 
//                         searcha a hero
//                     </div>
//                 }
//                 {
//                     (q !== "" && heroesFiltered.length === 0) &&
//                     <div className = "alert alert-danger"> 
//                         there is not a hero with {q} name.
//                     </div>
//                 }



//                 {
//                     heroesFiltered.map(hero => (
//                         <HeroCard
//                             ket = {hero.id}
//                             {...hero}

//                         />
//                     ))
//                 }


//                 </div>
//             </div>

//         </div>
//     )
// }
