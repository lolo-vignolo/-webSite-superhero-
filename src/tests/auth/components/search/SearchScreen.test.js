import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { SearchScreen } from "../../../../components/search/SearchScreen"

describe('teste de mi componente <SearchScreen />', () => {
 
    // const history = {
    //     length : 10,
    //     push : jest.fn(),
    //     goBack : jest.fn(),
    //     location:{}
    //}
    test('should show a default screen', () => {
        
        const wrapper = mount (
            <MemoryRouter initialEntries = {["/search"]}>

                <Route  path = "/search" component={SearchScreen}/>
              
            </MemoryRouter>
        
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe("searcha a hero")
    })

    test('should show to Batman and an imput with a value of  queryScreen', () => {
        
        const wrapper = mount (
            <MemoryRouter initialEntries = {["/search?q=batman"]}>

                <Route  path = "/search" component={SearchScreen}/>
              
            </MemoryRouter>
        
        )

        
        expect(wrapper.find("input").prop("value")).toBe("batman");
        expect(wrapper).toMatchSnapshot();
    })

    test('should show an error if the hero is not found', () => {
        const q = "batman!!!"
        const wrapper = mount (
            <MemoryRouter initialEntries = {[`/search?q=${q}`]}>

                <Route  path = "/search" component={SearchScreen}/>
              
            </MemoryRouter>
        
        )

        
        expect(wrapper.find(".alert-danger").text().trim()).toBe(`there is not a hero with ${q} name.`);
       
    })

    test('should calls push from history ', () => {
        const history = {
            push : jest.fn(),
        };

        const wrapper = mount (
            <MemoryRouter initialEntries = {[`/search?q=${q}`]}>

                <Route  path = "/search" component={()=> <SearchScreen history = {history} />}/>
              
            </MemoryRouter>
        
        )

        // diferencia entre simulate y props
        // => el simulate es mas recomentable cuando quiero escribir algo o interactuar, por ejemplo en el imput.
        // => el props es mas util, cuando solo necesito llamarlo.
        
        //a) aqui hago el cambio en la caja de texto.
        wrapper.find("imput").simulate("change",{
            target:{
                name:"superhero",
                value: "batman"
            }
        });

        // b ) aqui hacemos el submit del formulario
        wrapper.find("form").prop("onSubmit")({
            preventDefault(){}
        })


        //aqui esta cambiando el URL, el push en este caso cambia el URL.
        expect(history.push).toHaveBeenCalledWith("?q=batman")
    })
    


    
    
    
})
