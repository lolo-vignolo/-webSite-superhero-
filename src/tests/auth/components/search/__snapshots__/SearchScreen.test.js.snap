// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`teste de mi componente <SearchScreen /> should show a default screen 1`] = `
<div>
  <h1>
    Search Screen 
  </h1>
  <hr />
  <div
    className=" row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className=" form-control"
          name="superhero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-outline-primary btn-block"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        searcha a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`teste de mi componente <SearchScreen /> should show to Batman and an imput with a value of  queryScreen 1`] = `
<div>
  <h1>
    Search Screen 
  </h1>
  <hr />
  <div
    className=" row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className=" form-control"
          name="superhero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-outline-primary btn-block"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
                 
              </h5>
              <p
                className="card-text"
              >
                 
                Bruce Wayne
                 
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #27
                </small>
              </p>
              <a
                className="link-secondary"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
